channels/rdpgfx/server/rdpgfx_main.c

```diff
@@ -0,0 +1,1659 @@
+	Stream_Write_UINT8(s, pdu->pixelFormat); /* RDPGFX_PIXELFORMAT (1 byte) */
+	return rdpgfx_server_single_packet_send(context, s);
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+UINT rdpgfx_send_delete_surface_pdu(RdpgfxServerContext* context,
+									RDPGFX_DELETE_SURFACE_PDU* pdu)
+{
+	wStream* s = rdpgfx_server_single_packet_new(
+					 RDPGFX_CMDID_DELETESURFACE,
+					 sizeof(RDPGFX_DELETE_SURFACE_PDU));
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "rdpgfx_server_single_packet_new failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	Stream_Write_UINT16(s, pdu->surfaceId); /* surfaceId (2 bytes) */
+	return rdpgfx_server_single_packet_send(context, s);
+}
+
+static INLINE void rdpgfx_write_start_frame_pdu(wStream* s,
+		RDPGFX_START_FRAME_PDU* pdu)
+{
+	Stream_Write_UINT32(s, pdu->timestamp); /* timestamp (4 bytes) */
+	Stream_Write_UINT32(s, pdu->frameId); /* frameId (4 bytes) */
+}
+
+static INLINE void rdpgfx_write_end_frame_pdu(wStream* s,
+		RDPGFX_END_FRAME_PDU* pdu)
+{
+	Stream_Write_UINT32(s, pdu->frameId); /* frameId (4 bytes) */
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_send_start_frame_pdu(RdpgfxServerContext* context,
+										RDPGFX_START_FRAME_PDU* pdu)
+{
+	wStream* s = rdpgfx_server_single_packet_new(
+					 RDPGFX_CMDID_STARTFRAME,
+					 sizeof(RDPGFX_START_FRAME_PDU));
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "rdpgfx_server_single_packet_new failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	rdpgfx_write_start_frame_pdu(s, pdu);
+	return rdpgfx_server_single_packet_send(context, s);
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_send_end_frame_pdu(RdpgfxServerContext* context,
+									  RDPGFX_END_FRAME_PDU* pdu)
+{
+	wStream* s = rdpgfx_server_single_packet_new(
+					 RDPGFX_CMDID_ENDFRAME,
+					 sizeof(RDPGFX_END_FRAME_PDU));
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "rdpgfx_server_single_packet_new failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	rdpgfx_write_end_frame_pdu(s, pdu);
+	return rdpgfx_server_single_packet_send(context, s);
+}
+
+/**
+ * Function description
+ * Estimate RFX_AVC420_BITMAP_STREAM structure size in stream
+ *
+ * @return estimated size
+ */
+static INLINE UINT32 rdpgfx_estimate_h264_avc420(
+	RDPGFX_AVC420_BITMAP_STREAM* havc420)
+{
+	return sizeof(UINT32) /* numRegionRects */
+		   + (sizeof(RECTANGLE_16) + 2) /* regionRects + quantQualityVals */
+		   * havc420->meta.numRegionRects
+		   + havc420->length;
+}
+
+/**
+ * Function description
+ * Estimate surface command packet size in stream without header
+ *
+ * @return estimated size
+ */
+static INLINE UINT32 rdpgfx_estimate_surface_command(RDPGFX_SURFACE_COMMAND* cmd)
+{
+	RDPGFX_AVC420_BITMAP_STREAM* havc420 = NULL;
+	RDPGFX_AVC444_BITMAP_STREAM* havc444 = NULL;
+	UINT32 h264Size = 0;
+
+	/* Create new stream according to codec. */
```

Comment:
Not accurate.... It's actually totally wrong :)
I should have check the comments when I split code into different functions